### PR TITLE
feat(coinjoin): base classes

### PR DESCRIPTION
**Pull request by Syn-McJ**
## Issue being fixed or feature implemented
Some CoinJoin base classes need to be written, roughly by analogy with https://github.com/dashpay/dashj/pull/181/commits/22b666ae84ef1c78a2ad848d6f316c96250adf9e#diff-208066db8a4a90733fc6c3ca367b5ecfae162a975d4e9cfaeb99113ae4151226

## What was done?
A first draft of the following classes (many unknowns still remain):
- CoinJoin
- CoinJoinBaseSession
- Other related changes
- Some tests

Some p2p logic will be delegated back to DashSync through callback, which will be a part of the next PR.

## How Has This Been Tested?
N/A

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

### CODE REVIEW

File: dash-spv-apple-bindings/src/coinjoin.rs [hunk 1]
```diff
@@ -2,9 +2,24 @@ use core::slice;
...
 use dash_spv_coinjoin::messages;
+use dash_spv_coinjoin::coinjoin::CoinJoin;
+use dash_spv_coinjoin::callbacks::{GetInputValueByPrevoutHash, HasChainLock};
 use dash_spv_masternode_processor::consensus::Decodable;
 use ferment_interfaces::boxed;
 
+#[no_mangle]
+pub unsafe extern "C" fn register_coinjoin(
+    get_input_value_by_prevout_hash: GetInputValueByPrevoutHash,
+    has_chain_lock: HasChainLock,
```
Comment (Syn-McJ) [inline dash-spv-apple-bindings/src/coinjoin.rs:13]: I assume here DashSync will have some callbacks for p2p functions implemented.

File: dash-spv-coinjoin/src/coinjoin.rs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+
+#[repr(C)]
+#[derive(Debug)]
+// #[ferment_macro::export]
+pub struct CoinJoin {
+    pub opaque_context: *const std::ffi::c_void,
+    // pub chain_type: ChainType,
+    pub get_input_value_by_prevout_hash: GetInputValueByPrevoutHash,
+    pub has_chain_lock: HasChainLock,
+    map_dstx: HashMap<UInt256, CoinJoinBroadcastTx> // TODO: thread safety?
```
Comment (Syn-McJ) [inline dash-spv-coinjoin/src/coinjoin.rs:28]: Not sure what thread safety will look like yet. Don't see `mutex` being used in dash_shared_core

Comment (pankcuf) [reply dash-spv-coinjoin/src/coinjoin.rs:26]: At this point there no thread management in dash-shared-core. It's controlled by DashSync

**Review by pankcuf — APPROVED** [review]
At this point there is no multi-thread management at dash-shared-core. It's up to DashSync

**Pull request merged**